### PR TITLE
feat(atenlib): scaled_dot_product_attention; trilu

### LINKED ISSUE #547: aten::scaled_dot_product_attention
https://github.com/pytorch/pytorch/issues/97262

### PR DESCRIPTION
**Pull request by justinchuby**
Fixes #547 

Thanks @kunal-vaishnavi for helping me debug

What I learned:

1. We need SwapAxes
2. Being able to print in functions is absolutely saving time

### CODE REVIEW

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 1]
```diff
@@ -1073,6 +1074,184 @@ def aten_rrelu_with_noise_backward(
...
+    query: TFloat,
+    key: TFloat,
+    value: TFloat,
+    attn_mask: Optional[BOOL] = None,
+    dropout_p: float = 0.0,
+    is_causal: bool = False,
+    scale: Optional[float] = None,
+):
+    """scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False, *, float? scale=None) -> Tensor"""
+    # Use trace_only to obtain the permutation for Transpose
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/nn.py:1111]:
```suggestion
```

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 2]
```diff
@@ -1073,6 +1074,184 @@ def aten_rrelu_with_noise_backward(
...
+    query: TFloat,
+    key: TFloat,
+    value: TFloat,
+    attn_mask: Optional[TFloat] = None,
+    dropout_p: float = 0.0,
+    is_causal: bool = False,
+    scale: Optional[float] = None,
+):
+    """scaled_dot_product_attention(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False, *, float? scale=None) -> Tensor"""
+    # Use trace_only to obtain the permutation for Transpose
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/nn.py:1205]:
```suggestion
    # Use trace_only to handle optional inputs
```

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 3]
```diff
@@ -1073,6 +1074,208 @@ def aten_rrelu_with_noise_backward(
     raise NotImplementedError()
 
 
+@onnxscript.script()
+def _causal_attention_mask(query: TFloat, key: TFloat) -> TFloat:
+    # attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
+    # attn_mask = attn_mask.masked_fill(not attn_mask, -float('inf')) if attn_mask.dtype==torch.bool
+    target_length = op.Shape(query)[-2:-1]
+    source_length = op.Shape(key)[-2:-1]
+    # attn_mask = torch.ones(L, S) := {
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_aten/ops/nn.py:1083]: For my learning, what is this style of comment?

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 4]
```diff
@@ -1073,6 +1074,208 @@ def aten_rrelu_with_noise_backward(
     raise NotImplementedError()
 
 
+@onnxscript.script()
+def _causal_attention_mask(query: TFloat, key: TFloat) -> TFloat:
+    # attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_aten/ops/nn.py:1079]: Consider putting in docstring to make it obvious that these two lines are the torch reference implementation of this function?

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 5]
```diff
@@ -1073,6 +1074,208 @@ def aten_rrelu_with_noise_backward(
...
+):
+    # Swap the last two axes of key
+    key_shape = op.Shape(key)
+    key_last_dim = key_shape[-1:]
+    key_second_last_dim = key_shape[-2:-1]
+    key_first_dims = key_shape[:-2]
+    key_squeezed_shape = op.Concat(
+        op.Constant(value_ints=[-1]), key_second_last_dim, key_last_dim, axis=0
+    )
+    key_squeezed = op.Reshape(key, key_squeezed_shape)
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_aten/ops/nn.py:1262]: I think you can do `Reshape(key, op.Constant(value_ints=[-1, 0, 0]))`

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 4 above)
Comment (BowenBao) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1079]: what is `L` and `S`?

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 6]
```diff
@@ -1073,6 +1074,208 @@ def aten_rrelu_with_noise_backward(
...
+    # }
+    attn_mask = op.Trilu(attn_mask, upper=0)
+    # The causal mask has 0s in the lower triangle and -inf in the upper triangle.
+    attn_mask = op.Where(op.Equal(attn_mask, 0.0), op.Constant(value_float=-float("inf")), 0.0)
+    return attn_mask
+
+
+@onnxscript.script()
+def _attention_scale(query: TFloat) -> TFloat:
+    # scale_factor = 1 / math.sqrt(Q.size(-1))
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_aten/ops/nn.py:1095]: nit: Q, K, V are pretty obvious, but wonder if it is still better to just write as `query` in comment

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 7]
```diff
@@ -1073,6 +1074,208 @@ def aten_rrelu_with_noise_backward(
...
+
+    # https://github.com/pytorch/pytorch/blob/12da0c70378b5be9135c6fda62a9863bce4a4818/aten/src/ATen/native/transformers/attention.cpp#L653
+    # Scale q, k before matmul for stability see https://tinyurl.com/sudb9s96 for math
+    query_scaled = op.Mul(query, op.Sqrt(scale))
+    key_transposed_scaled = op.Mul(key_transposed, op.Sqrt(scale))
+    attn_weight = op.Softmax(
+        op.MatMul(query_scaled, key_transposed_scaled),
+        axis=-1,
+    )
+    attn_weight, _ = op.Dropout(attn_weight, dropout_p)
```
Comment (BowenBao) [inline onnxscript/function_libs/torch_aten/ops/nn.py:1207]: How does dropout work in export in training vs inference?

**Review by BowenBao — COMMENTED** [review]
🔥 🔥 🔥 This is great!

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 3 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1083]: it was my own style :D lmk if they should look differently

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 5 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1262]: Nice!

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 3 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1083]: To show that the lines are equivalent to the torch op in the comment

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 8]
```diff
@@ -1073,6 +1074,208 @@ def aten_rrelu_with_noise_backward(
...
+            query, key, value, scale, dropout_p
+        )
+
+    return _aten_scaled_dot_product_attention_bool_mask_onnx(
+        query, key, value, attn_mask, scale, dropout_p
+    )
+
+
+@torch_op("aten::scaled_dot_product_attention", trace_only=True, overload=True)
+def aten_scaled_dot_product_attention_float_mask(
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/nn.py:1146]: From the looks of the tests I think this overload should be the default

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 7 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1207]: I think it keeps the same behavior. @kunal-vaishnavi Is that right

**Comment by justinchuby** [comment]
ORT crashes on Windows (sometimes)

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 4 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1079]: L is the target sequence length, S is the source sequence length, and E is the embedding size. Added a comment

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 6 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/nn.py]: Done

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 5 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1293]: https://github.com/microsoft/onnxruntime/issues/15203

**Comment by BowenBao** [comment]
> 2. Being able to print in functions is absolutely saving time

Would it be helpful to author functions in eager mode, where print is available, then add the `@script` in the end?

🤔 Do you anticipate a lot of friction in the last step?

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 9]
```diff
@@ -863,6 +879,28 @@ def _where_input_wrangler(
...
     ),
+    skip(
+        "nn.functional.scaled_dot_product_attention",
+        matcher=lambda sample: (attn_mask := sample.kwargs.get("attn_mask")) is not None
+        and attn_mask.dtype == torch.bool,
+        reason="this overload takes a non-boolean mask",
+    ),
+    skip(
+        "nn.functional.scaled_dot_product_attention",
+        matcher=lambda sample: sample.kwargs.get("dropout_p") != 0.0,
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:890]: dropout will behave the same way eval or training. Ops are all functional so if an op has a training behavior it will have an argument controlling that behavior, I believe.

[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:890]: Let me check dropout

[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:890]: Yeah: for example aten::dropout itself has `dropout(Tensor input, float p, bool train) -> Tensor`

[the remark this answers is not in the document]
Comment (BowenBao) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:890]: It appears `scaled_dot_product_attention` is used internally and has a more functional interface. The `dropout_p` is set by caller (e.g., `multi_head_attention_forward`, multi_head_attention_forward) depending on training mode.

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 7 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1234]: That makes sense. But in any case this function is a pure function that is controlled only by its input

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (justinchuby) [reply onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:890]: Nice!

File: onnxscript/function_libs/torch_aten/ops/nn.py (hunk 7 above)
Comment (BowenBao) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1234]: Yes

Comment (BowenBao) [reply onnxscript/function_libs/torch_aten/ops/nn.py:1234]: > How does dropout work in export in training vs inference?

To answer the original question. For this operator, `training` vs `inference` is set for the model before calling export. The mode affects and alters the fx graph. No action needed from exporter.

E.g., If model is in inference mode, exporter receives `aten::scaled_dot_product_attention` with `dropout_p=0.0`. And `dropout_p=<some_preset_value>` if in training mode.

**Review by BowenBao — APPROVED** [review]
🚢

**Comment by justinchuby** [comment]
>ORT crashes on Windows (sometimes)

Ubuntu with onnx-weekly also crashed. Hmm

We need to debug this in ort for errors

```
2023-03-27T17:39:13.6127530Z [31m[1m______ onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py _______[0m
2023-03-27T17:39:13.6179068Z [gw0] linux -- Python 3.10.10 /home/runner/work/onnx-script/onnx-script/.nox/test_onnx_weekly/bin/python
2023-03-27T17:39:13.6180003Z worker 'gw0' crashed while running 'onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py::TestOutputConsistencyFullGraphCPU::test_output_match_opinfo__nn_functional_scaled_dot_product_attention_bool_mask_cpu_float32'
```

For awareness @gramalingam @liqunfu @jcwchen

**Comment by xuzheyuan624** [comment]
How can I regisiter this ops in my code？
